js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,362 @@
+			socket.handleEvent("error", rt.ToValue(readErr))
+
+		case scheduledFn := <-socket.scheduled:
+			scheduledFn(goja.Undefined())
+
+		case <-ctx.Done():
+			// This means that K6 is shutting down (e.g., during an interrupt)
+			socket.handleEvent("close", rt.ToValue("Interrupt"))
+			socket.closeConnection(websocket.CloseGoingAway)
+
+		case <-socket.done:
+			// This is the final exit point normally triggered by closeConnection
+			end := time.Now()
+			sessionDuration := stats.D(end.Sub(start))
+
+			samples := []stats.Sample{
+				{Metric: metrics.WSSessions, Time: end, Tags: tags, Value: 1},
+				{Metric: metrics.WSConnecting, Time: end, Tags: tags, Value: connectionDuration},
+				{Metric: metrics.WSSessionDuration, Time: end, Tags: tags, Value: sessionDuration},
+				{Metric: metrics.DataReceived, Time: end, Tags: tags, Value: float64(bytesRead)},
+				{Metric: metrics.DataSent, Time: end, Tags: tags, Value: float64(bytesWritten)},
+			}
+			state.Samples = append(state.Samples, samples...)
+
+			return response
+		}
+	}
+}
+
+func (s *Socket) On(event string, handler goja.Value) {
+	if handler, ok := goja.AssertFunction(handler); ok {
+		s.eventHandlers[event] = append(s.eventHandlers[event], handler)
+	}
+}
+
+func (s *Socket) handleEvent(event string, args ...goja.Value) {
+	if handlers, ok := s.eventHandlers[event]; ok {
+		for _, handler := range handlers {
+			handler(goja.Undefined(), args...)
+		}
+	}
+}
+
+func (s *Socket) Send(message string) {
+	// NOTE: No binary message support for the time being since goja doesn't
+	// support typed arrays.
+	rt := common.GetRuntime(s.ctx)
+
+	writeData := []byte(message)
+	if err := s.conn.WriteMessage(websocket.TextMessage, writeData); err != nil {
+		s.handleEvent("error", rt.ToValue(err))
+	}
+}
+
+func (s *Socket) Ping() {
+	rt := common.GetRuntime(s.ctx)
+	deadline := time.Now().Add(writeWait)
+
+	err := s.conn.WriteControl(websocket.PingMessage, []byte{}, deadline)
+	if err != nil {
+		s.handleEvent("error", rt.ToValue(err))
+	}
+}
+
+func (s *Socket) SetTimeout(fn goja.Callable, timeoutMs int) {
+	// Starts a goroutine, blocks once on the timeout and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		select {
+		case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
+			s.scheduled <- fn
+
+		case <-s.done:
+			return
+		}
+	}()
+}
+
+func (s *Socket) SetInterval(fn goja.Callable, intervalMs int) {
+	// Starts a goroutine, blocks forever on the ticker and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		ticker := time.NewTicker(time.Duration(intervalMs) * time.Millisecond)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.scheduled <- fn
+
+			case <-s.done:
+				return
+			}
+		}
+	}()
+}
+
+func (s *Socket) Close(args ...goja.Value) {
+	code := websocket.CloseGoingAway
+	if len(args) > 0 && !goja.IsUndefined(args[0]) && !goja.IsNull(args[0]) {
```

Comment:
I haven't checked this, but... do you need to check for undefined/null here? I thought {null}.ToInteger() and {undefined}.ToInteger() just returned 0? (I may be wrong.)